larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h

```diff
@@ -0,0 +1,393 @@
+/**
+ *  @file   larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h
+ *
+ *  @brief  Header file for the lar discrete probability vector class
+ *
+ *  $Log: $
+ */
+#ifndef LAR_DISCRETE_PROBABILITY_VECTOR_H
+#define LAR_DISCRETE_PROBABILITY_VECTOR_H 1
+
+#include "Pandora/StatusCodes.h"
+
+#include <limits>
+#include <random>
+#include <utility>
+#include <vector>
+
+namespace lar_content
+{
+
+/**
+ *  @brief  DiscreteProbabilityVector class
+ */
+class DiscreteProbabilityVector
+{
+public:
+
+    template <typename TX, typename TY>
+    using InputDatum = std::pair<TX, TY>;
+
+    template <typename TX, typename TY>
+    using InputData = std::vector<InputDatum<TX, TY> >;
+
+    typedef std::vector<float> ResamplingPoints;
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  inputData the data used to construt the probability vector
+     *  @param  xUpperBound the upper bound of the probability vector
+     *  @param  useWidths bool controlling whether the bin widths are used in calculations
+     */
+    template <typename TX, typename TY>
+    DiscreteProbabilityVector(const InputData<TX, TY> &inputData, const TX xUpperBound, const bool useWidths);
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  discreteProbabilityVector a discrete probability vector to resample from
+     *  @param  resamplingPoints the points to resample the discrete probability vector with 
+     */
+    DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector, const ResamplingPoints &resamplingPoints);
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  discreteProbabilityVector a discrete probability vector to randomly rearrange
+     *  @param  randomNumberGenerator the random number generator for the random reshuffling
+     */
+    DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector, std::mt19937 &randomNumberGenerator);
```

Comment:
Presumably RNG can't be const?

Reply:
Sadly it cannot.  The random throwing updates the internal state (which makes sense; otherwise you'd throw the same number repeatedly).